src/chapter4/openmp.md

```diff
@@ -1 +1,101 @@
-# OpenMP
+# Parallel Computing with OpenMP
+
+## What is OpenMP
+
+OpenMP, stand for open multi-processing is an API for writing multithreaded applications
+
+It has a set of compiler directives and library routines for parallel applications, and it greatly simplifies writing multi-threaded code in Fortran, C and C++.
+
+Just few lines of additional code can make your application parallel 
+
+OpenMP uses shared memory architecture. It assumes all code runs on a single server
+
+## Threads
+
+![Threads Visualisation](imgs/Threads%20Visualisation.png)
+
+A thread of execution is the smallest instruction that can be managed independently by an operating system.
+
+In parallel region, multiple threads are spawned and utilises the cores on CPU
+
+> Only one thread exists in a serial region
+
+## Compiler Directive \# pragma
+
+-   Source codes with “#” provide additional information to compiler
+-   `#define X 10`
+-   `#include <omp.h>`
+-   `#pragma omp parallel`
+  
+OpenMP provides a set of `#pragma` directives that can be used to specify the parallelization of a particular loop or section of code. For example, the `#pragma omp parallel` directive is used to start a parallel region, where multiple threads can execute the code concurrently. The `#pragma omp for` directive is used to parallelize a loop, with each iteration of the loop being executed by a different thread.
+
+Here's an example of how `#pragma` directives can be used with OpenMP to parallelize a simple loop:
+  
+  
+Use `gcc -fopenmp` to compile your code when you use `#pragma`
+
+## Compile OpenMP
+
+1. Add `#include <omp.h> if you are using OpenMP function`
+2. Run `gcc -fopenmp -o hello hello.c`
+
+## How it works
+
+![OpenMP and Directive](imgs/OpenMP%20and%20Directive.png)
+[Source](https://www.researchgate.net/figure/OpenMP-API-The-master-thread-is-indicated-with-T-0-while-inside-the-parallel-region_fig3_329536624 
+)
+
+Here is an example of `#pragma`
+- The function starts with serial region
+- At the line `#pragma omp parallel`, a group of threads are spawned to create parallel region inside the bracket 
+- At the end of the bracket, the program goes back to serial computing 
+
+## Running "Hello World" on Multi-threads
+
+>If you're unsure about the difference between **multi-threading** and **multi-processing**, check the page [here](src/chapter4/multithreading.md)
```

Comment:
change the link to a relative one, eg `./multithreading.md`

Reply:
Addressed